src/dune_pkg/rev_store.ml

```diff
@@ -407,18 +407,40 @@ let query_head_branch =
       | Some m -> Some (Re.Group.get m 1))
 ;;
 
-let read_head_branch =
-  let headline = Re.(compile @@ seq [ bol; rep space; str "Remote branch:" ]) in
-  fun t handle ->
-    let+ lines = run_capture_lines t [ "remote"; "show"; "-n"; handle ] in
-    let rec inspect = function
-      | [] | [ _ ] -> None
-      | heading :: branch :: rest ->
-        (match Re.exec_opt headline heading with
-         | None -> inspect (branch :: rest)
-         | Some _ -> Some (String.trim branch))
-    in
-    inspect lines
+let branch_of_refspec refspec =
+  refspec
+  |> String.drop_prefix_if_exists ~prefix:"+"
+  |> String.lsplit2 ~on:':'
+  |> Option.bind ~f:(fun (remote_ref, _local_ref) ->
+    String.drop_prefix remote_ref ~prefix:"refs/heads/")
+;;
+
+let read_head_branch t handle =
+  let headline =
+    Re.(compile @@ seq [ bol; str (sprintf {|[remote "%s"]|} handle); eol ])
+  in
+  let fetch_line = Re.(compile @@ seq [ str "fetch = "; group (rep1 any); eol ]) in
+  let path = Path.relative t.dir "config" in
+  let lines = Io.read_file path |> String.split ~on:'\n' in
+  let rec inspect_in_section = function
+    | [] -> None
+    | line :: lines ->
+      (match Re.exec_opt fetch_line line with
+       | None -> inspect_in_section lines
+       | Some m ->
+         let refspec = Re.Group.get m 1 in
+         (match branch_of_refspec refspec with
+          | Some b -> Some b
+          | None -> inspect_in_section lines))
+  in
+  let rec inspect = function
```

Comment:
I think this is just `List.find_map`?

Reply:
It is not, as it pops of lines from the list until the heading is found. But turns out it is fairly similar to `List.split_while`, so I rewrote it to use that.